File: DemoServer/AdditionalLanguages/python/staticIndexes/projectIndexResults/projectIndexResults.py
from demo_example import Example, RunParamsBase

#region Usings
from ravendb import (
    AbstractIndexCreationTask,
    QueryData
)
#endregion

#region Demo
#region Step_1
class Employees_ByWorkPeriod(AbstractIndexCreationTask):
#endregion

    #region Step_2
    class IndexEntry:
        def __init__(self, working_in_company_since: int = None):
            self.working_in_company_since = working_in_company_since
    #endregion

    #region Step_3
    class EmployeeProjectedDetails:
        def __init__(self, first_name: str = None, phone: str = None, location: str = None):
            self.first_name = first_name
            self.phone = phone
            self.location = location
    #endregion

    #region Step_4
    def __init__(self):
        super().__init__()
        self.map = (
            "docs.Employees.Select(employee => new { "
            "    working_in_company_since = employee.HiredAt.Year "
            "})"
        )
    #endregion
#endregion

class RunParams(RunParamsBase):
    def __init__(self, start_year: int = None):
        self.start_year = start_year

class ProjectIndexResults(Example):
    def run(self, run_params: RunParams):
        startYear = run_params.start_year or 1993
        Employees_ByWorkPeriod().execute(self.document_store_holder.store())
        #region Demo
        with self.document_store_holder.store().open_session() as session:
            #region Step_5
            employees_query = session.query_index_type(
                Employees_ByWorkPeriod, Employees_ByWorkPeriod.IndexEntry
            ).where_greater_than("working_in_company_since", startYear)
            #endregion
            
            #region Step_6
            employees_query = employees_query.select_fields_query_data(
                Employees_ByWorkPeriod.EmployeeProjectedDetails,
                QueryData.custom_function(
                    "employee ",
                    (
                        "{ first_name: employee.FirstName, "
                        " phone: employee.HomePhone, "
                        " location: employee.Address.City + ' ' + employee.Address.Country } "
                    ),
                ),
            )
            #endregion

            #region Step_7
            employees_since_year = list(employees_query)
            #endregion
        #endregion
        return employees_since_year

File: DemoServer/AdditionalLanguages/php/staticIndexes/projectIndexResults/ProjectIndexResults.php
<?php

namespace RavenDB\Demo\staticIndexes\projectIndexResults;

//region Usings
use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Documents\Queries\QueryData;
//endregion

use RavenDB\Demo\common\DocumentStoreHolder;

//region Demo
//region Step_1
class Employees_ByWorkPeriod extends AbstractIndexCreationTask
//endregion
{
    //region Step_2
    public function __construct()
    {
        parent::__construct();
        $this->map =
            "docs.Employees.Select(employee => new {" .
            "      WorkingInCompanySince = employee.HiredAt.Year" .
            "})";
    }
    //endregion
}
//endregion

//region Demo
//region Step_3
class IndexEntry
{
    private ?int $workingInCompanySince = null;

    public function getWorkingInCompanySince(): ?int
    {
        return $this->workingInCompanySince;
    }

    public function setWorkingInCompanySince(?int $workingInCompanySince): void
    {
        $this->workingInCompanySince = $workingInCompanySince;
    }
}
//endregion
//endregion

//region Demo
//region Step_4
class EmployeeProjectedDetails
{
    private ?string $firstName = null;
    private ?string $phone = null;
    private ?string $location = null;

    public function getFirstName(): ?string
    {
        return $this->firstName;
    }

    public function setFirstName(?string $firstName): void
    {
        $this->firstName = $firstName;
    }

    public function getPhone(): ?string
    {
        return $this->phone;
    }

    public function setPhone(?string $phone): void
    {
        $this->phone = $phone;
    }

    public function getLocation(): ?string
    {
        return $this->location;
    }

    public function setLocation(?string $location): void
    {
        $this->location = $location;
    }
}
//endregion
//endregion

class ProjectIndexResults
{
    public function __invoke(RunParams $runParams): array
    {
        $startYear = $runParams->getStartYear();

        //region Demo
        $employeesSinceYear = [];

        $session = DocumentStoreHolder::getStore()->openSession();
        
        try {
            $employeesQuery = $session
                //region Step_5
                ->query(IndexEntry::class, Employees_ByWorkPeriod::class)
                ->whereGreaterThan("WorkingInCompanySince", $startYear)
                //endregion
                //region Step_6
                ->selectFields(EmployeeProjectedDetails::class,
                    QueryData::customFunction("employee ",
                        "{ FirstName: employee.FirstName," .
                        "  Phone: employee.HomePhone," .
                        "  Location: employee.Address.City + ' ' + employee.Address.Country }"));
                //endregion

            //region Step_7
            $employeesSinceYear = $employeesQuery->toList();
            //endregion
            
        } finally {
            $session->close();
        }
        //endregion

        return $employeesSinceYear;
    }
}
